app/services/ui_notifications.rb

```diff
@@ -0,0 +1,31 @@
+# Namespace for Foreman UI notifications event handling
+module UiNotifications
+  class Base
+    attr_reader :state, :subject
+
+    def initialize(subject, state)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
+      @subject = subject
+      raise(Foreman::Exception, 'invalid state') unless allowed_states.include?(state)
```

Comment:
https://github.com/theforeman/foreman/pull/4339/files#r103540123 , also if you want to validate, how about putting all validations at the beginning? That way we don't set `@subject = subject` for nothing in the case of an invalid `state`